### PR TITLE
Adress zero index tl issue

### PR DESCRIPTION
**Pull request by SharkyZg**
When the car reaches end of the track in the simulator waypoint index resets to 0. That would cause problems for traffic light detection before this fix because traffic lights are detected based on difference between waypoint index of car and waypoint index of traffic light.

I have addressed this issue and exported closest traffic light detection code to another get_closest_tl_wp_idx function to make the code more readable.

### CODE REVIEW

**Comment by SharkyZg** [comment]
Please squash and merge commits if you decide to merge.

**Comment by carlosgalvezp** [comment]
Thanks! I'll have a look tomorrow :)

File: ros/src/tl_detector/tl_detector.py [hunk 1]
```diff
@@ -269,18 +269,63 @@ def get_light_state(self, light):
...
+        # issue
+        last_waypoint_index = len(self.waypoints.waypoints) - 1
+        car_distance_from_last_wp = last_waypoint_index - car_pos_wp_idx
+        # Loop thorugh waypoints to find the closest traffic ligt waypoint
+        smallest_tl_distance = 10000
+        for tl_waypoint_idx in self.tl_waypoints_idx:
+            # Cover corner case where traffic light is just behind the
+            # waypoint 0 And the car is near last waypoint so waypoint
+            # index will reset to 0
+            if(searching_dist_tl > (car_distance_from_last_wp) and
```
Comment (idanov) [inline ros/src/tl_detector/tl_detector.py]: Here the 80 characters line limit makes the code hard to read... Could we bump up the limit to 120 characters? @carlosgalvezp

File: ros/src/tl_detector/tl_detector.py [hunk 2]
```diff
@@ -293,33 +338,25 @@ def process_traffic_lights(self):
...
+                )
 
         # If waypoint has been found get traffic light state
-        if light_wp > -1:
-            light = self.waypoints.waypoints[light_wp]
+        if light_wp_idx > -1:
+            light = self.waypoints.waypoints[light_wp_idx]
             state = self.get_light_state(light)
-            return light_wp, state
+            print(light_wp_idx)
```
Comment (idanov) [inline ros/src/tl_detector/tl_detector.py]: Could we `rospy.loginfo()` here?

**Review by idanov — APPROVED** [review]
Great, LGTM! I didn't checkout and test it, but I assume it works as expected. I will test it tomorrow if you haven't merged it yet.

File: ros/src/tl_detector/tl_detector.py (hunk 2 above)
Comment (SharkyZg) [reply ros/src/tl_detector/tl_detector.py]: @idanov sure, or I can remove this line completely, I've forgot to remove it after testing.

File: ros/src/tl_detector/tl_detector.py [hunk 3]
```diff
@@ -268,18 +268,76 @@ def get_light_state(self, light):
...
+            light_wp_idx: waypoint index of closest traffic light in range
+
+        """
+
+        light_wp_idx = -1
+        # In case that array with traffic light waypoints does not exist,
+        # create it
+        if(self.tl_waypoints_idx == []):
+            self.get_tl_waypoints_idx()
+        last_waypoint_idx = len(self.waypoints.waypoints) - 1
```
Comment (carlosgalvezp) [inline ros/src/tl_detector/tl_detector.py]: When passed to the function `get_distance_in_track`, you should pass the actual length (as number of waypoints), not the index to the last element. So it should not have the `-1` there. Maybe you can call it `track_length` or similar as well?

Also, `car_distance_from_last_wp` is not used anymore so can be removed

**Review by carlosgalvezp — COMMENTED** [review]
Looks really good! Just a minor comment

**Review by carlosgalvezp — APPROVED** [review]

**Pull request merged**